snapcraft/internal/sources/errors.py

```diff
@@ -21,6 +21,28 @@ class VCSError(errors.SnapcraftError):
     fmt = '{message}'
 
 
+class SnapcraftSourceUnhandledError(errors.SnapcraftError):
+
+    fmt = ('Failed to pull source for part {part_name!r}: '
+           'unable to determine source type of {source!r}.\n'
+           'Check that the URL is correct or '
+           'consider specifying `source-type` for this part. '
+           'See `snapcraft help sources` for more information.')
+
+    def __init__(self, source, part_name):
+        super().__init__(source=source, part_name=part_name)
+
+
+class SnapcraftSourceNotADirectoryError(errors.SnapcraftError):
+
+    fmt = ('Failed to pull source for part {part_name!r}: '
+           '{source!r} is not a directory.\n'
```

Comment:
The exception name has been misleading the message, which I guess comes from the check (`isdir(source)`). The implications of that is that we found no handler to handle a file source (`source: mytar.tgz`) which is not a directory nor is this source a directory, maybe something like this would be better:
```
'Failed to pull source for part {part_name!r}: '
'the `source-type` for {source!r} could not be detected.\n'
```

Or change the logic a bit,
```python
     elif _tar_type_regex.match(source):
         source_type = 'tar'
     elif  os.path.is_dir(source):
        source_type = 'dir'
     else:
        raise errors.SnapcraftSourceUnhandledError(source, part_name)
```

and in the `_source_handler` add `'dir': Local`.

This rids us of that last exception.